src/dbnode/storage/block/lease.go

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package block
+
+import (
+	"errors"
+	"sync"
+)
+
+var (
+	errLeaserAlreadyRegistered = errors.New("leaser already registered")
+	errAddLeaseLeaserNotRegistered = errors.New("add lease error as leaser not registered")
+)
+
+type leaseManager struct {
+	sync.Mutex
+	leasers []Leaser
+	verifier LeaseVerifier
+}
+
+// NewLeaseManager creates a new lease manager with a provided
+// lease verifier (to ensure leases are valid when made).
+func NewLeaseManager(verifier LeaseVerifier) LeaseManager {
+	return &leaseManager{
+		verifier: verifier,
+	}
+}
+
+func (m *leaseManager) RegisterLeaser(leaser Leaser) error {
+	m.Lock()
+	defer m.Unlock()
+
+	for _, l := range m.leasers {
+		if l == leaser {
+			return errLeaserAlreadyRegistered
+		}
+	}
+
+	m.leasers = append(m.leasers, leaser)
+
+	return nil
+}
+
+func (m *leaseManager) OpenLease( 
+	leaser Leaser,
+	descriptor LeaseDescriptor, 
+	state LeaseState,
+) error {
+	// NB(r): Take exclusive lock so that upgrade leases can't be called
+	// while we are verifying a lease (racey)
+	m.Lock()
+	defer m.Unlock()
+
+	registered := false
+	for _, l := range m.leasers {
+		if l == leaser {
```

Comment:
Do you think this is good enough for determining equality? We've just have to be careful basically to not create a new leaser each time...the alternative would be to have an `Identity() string` method and use that for comparison but I guess if you messed this up it would show up right away in tests

Reply:
I've used this a few times, the iface check will make sure it's both same type and same pointer type (since iface is really just struct with the underlying type as an enum like value and the pointer value of the actual receiver).